src/main/java/org/springframework/cloud/bindings/boot/pem/PemCertificateParser.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2012-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.bindings.boot.pem;
+
+import java.io.ByteArrayInputStream;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Parser for X.509 certificates in PEM format.
+ *
+ * @author Scott Frederick
+ * @author Phillip Webb
+ */
+final class PemCertificateParser {
+
+	private static final String HEADER = "-+BEGIN\\s+.*CERTIFICATE[^-]*-+(?:\\s|\\r|\\n)+";
+
+	private static final String BASE64_TEXT = "([a-z0-9+/=\\r\\n]+)";
+
+	private static final String FOOTER = "-+END\\s+.*CERTIFICATE[^-]*-+";
+
+	private static final Pattern PATTERN = Pattern.compile(HEADER + BASE64_TEXT + FOOTER, Pattern.CASE_INSENSITIVE);
+
+	private PemCertificateParser() {
+	}
+
+	/**
+	 * Parse certificates from the specified string.
+	 * @param certificates the certificates to parse
+	 * @return the parsed certificates
+	 */
+	static X509Certificate[] parse(String certificates) {
+		if (certificates == null) {
+			return null;
+		}
+		CertificateFactory factory = getCertificateFactory();
+		List<X509Certificate> certs = new ArrayList<>();
+		readCertificates(certificates, factory, certs::add);
+		return (!certs.isEmpty()) ? certs.toArray(X509Certificate[]::new) : null;
```

Comment:
`certs.toArray` does not exist in Java 8.
and this module still supports Java 8.

You can rather use the `stream` api: `certs.stream().`

Reply:
I'm a bit puzzled here, but if you couldn't tell, I am not a Java person, so I am sure I'm missing something. From what I can tell, ArrayList.toArray() is supported in Java 8: https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#toArray--

I also don't see a `stream` method on ArrayList.

Reply:
> ArrayList.toArray() is supported in Java 8

yes it is; but you declared `certs` as a `List` (not `ArrayList` which is concrete type), which inherits from `Collection` which only recently (well, Java `11`) added `toArray(IntFunction<T[]> generator)(IntFunction<T[]> generator` (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collection.html#toArray(java.util.function.IntFunction))

> I also don't see a stream method on ArrayList.
yes, it's on `Collection` as well (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collection.html#stream())

this compiles fine on Java 8:

```java
return (!certs.isEmpty()) ? certs.stream().toArray(X509Certificate[]::new) : null;
```